includes/Core/Assets/Assets.php

```diff
@@ -636,6 +654,37 @@ private function add_async_defer_attribute( $tag, $handle ) {
 		return $tag;
 	}
 
+	/**
+	 * Executes all extra callbacks before printing a list of dependencies.
+	 *
+	 * This method ensures that such callbacks that run e.g. `wp_add_inline_script()` are executed just-in-time,
+	 * only when the asset is actually loaded in the current request.
+	 *
+	 * This method works recursively, also looking at dependencies, and supports both scripts and stylesheets.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param WP_Dependencies $dependencies WordPress dependencies class instance.
+	 * @param array           $handles      List of handles to run before print callbacks for.
+	 */
+	private function run_before_print_callbacks( WP_Dependencies $dependencies, array $handles ) {
+		foreach ( $handles as $handle ) {
+			if ( isset( $this->print_callbacks_done[ $handle ] ) ) {
+				continue;
+			}
+
+			if ( isset( $this->assets[ $handle ] ) ) {
+				$this->assets[ $handle ]->before_print();
+			}
+
+			if ( isset( $dependencies->registered[ $handle ] ) && is_array( $dependencies->registered[ $handle ]->deps ) ) {
+				$this->run_before_print_callbacks( $dependencies, $dependencies->registered[ $handle ]->deps );
+			}
+
+			$this->print_callbacks_done[ $handle ] = true;
```

Comment:
This seems like it should be set right after `before_print` is called above (or even right after the initial conditional that checks it), otherwise there's nothing stopping the callback from being invoked for the same handle multiple times if one of it's dependencies also depends on another script that has `before_print` logic.